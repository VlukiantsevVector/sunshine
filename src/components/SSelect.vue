<template>
  <s-menu v-model="show">
    <s-button
      style="text-align: left; min-width: 150px"
      slot="activator"
      @keydown="onKeyPress"
      v-bind="$attrs"
      v-on="$listeners"
    >
      {{ text }}
      <template slot="right">
        <s-svg name="dropdownArrow" />
      </template>
    </s-button>
    <div slot="content">
      <ul class="s-select__list">
        <li v-for="item in items"
          class="s-select__list__item"
          :key="item.key"
          :class="{'s-select__list__item--selected': item === value}"
        >
          <a href="#" @click.prevent="onClick(item)">{{ item[labelKey] }}</a>
        </li>
      </ul>
    </div>
  </s-menu>
</template>


<script lang="ts">
import Vue from 'vue';
import SButton from './SButton.vue';
import SMenu from './SMenu.vue';
import SSvg from './internal/SSvg.vue';

export default Vue.extend({
  name: 's-select',
  inheritAttrs: false,

  components: {
    SButton,
    SMenu,
    SSvg,
  },

  props: {
    items: Array as () => any[],
    value: Object,
    labelKey: {
      type: String,
      default: 'title',
    },
    label: {
      type: String,
      default: 'Select',
    },
  },

  data() {
    return {
      show: false,
    };
  },

  computed: {
    text(): string {
      return this.value ? this.value[this.labelKey] : this.label;
    },

    currentIndex(): number {
      return this.items.indexOf(this.value);
    },
  },

  methods: {
    onClick(item: any) {
      this.setSelectedItem(item);
      this.show = false;
    },

    setSelectedItem(item: any) {
      this.$emit('input', item);
    },

    setNextSelected() {
      if (this.currentIndex < this.items.length - 1)
        this.setSelectedItem(this.items[this.currentIndex + 1]);
    },

    setPrevSelected() {
      if (this.currentIndex > 0)
        this.setSelectedItem(this.items[this.currentIndex - 1]);
    },

    onKeyPress(event: KeyboardEvent) {
      console.log(event.key);
      let consumed = false;
      switch (event.key) {
        case 'ArrowDown':
          this.setNextSelected();
          break;
        case 'ArrowUp':
          this.setPrevSelected();
          break;
      }
      return true;
    },
  },
});
</script>
