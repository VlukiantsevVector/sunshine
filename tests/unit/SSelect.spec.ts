import { mount, config } from '@vue/test-utils';
import SSelect from '@/components/SSelect.vue';

config.logModifiedComponents = false;

describe('SSelect', () => {
  it('renders select box', () => {
    const wrapper = mount(SSelect, {
      propsData: {
        items: [
          {title: 'Norge'},
          {title: 'Sverige'},
          {title: 'England'},
        ],
        label: 'Choose country',
      },
    });
    expect(wrapper.html()).toMatchSnapshot();
  });
});
