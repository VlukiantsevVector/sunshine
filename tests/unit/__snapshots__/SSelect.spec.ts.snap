// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SSelect renders select box 1`] = `
<span class="s-menu"><span><button class="s-btn" style="text-align: left; min-width: 150px;"><span tabindex="-1" class="inline-flex s-btn__content s-btn__content--secondary"> <span class="flex-grow">
    Choose country
    </span> <span>â¯†</span></span></button></span>
<!----></span>
`;
